diff --git a/src/SolrConnector/SolrConnectorPluginBase.php b/src/SolrConnector/SolrConnectorPluginBase.php
index ecb8918a..3332e935 100644
--- a/src/SolrConnector/SolrConnectorPluginBase.php
+++ b/src/SolrConnector/SolrConnectorPluginBase.php
@@ -1179,6 +1179,11 @@ abstract class SolrConnectorPluginBase extends ConfigurablePluginBase implements
       return $array_data[basename($filepath)];
     }
 
+    // In case of file hosted on s3fs array_data has full file path as key.
+    if (isset($array_data[$filepath])) {
+      return $array_data[$filepath];
+    }
+
     // In most (or every) cases when an error happens we won't reach that point,
     // because a Solr exception is already pased through. Anyway, this exception
     // will be thrown if the solarium library surprises us again. ;-)
