diff --git a/modules/contrib/filefield_paths/filefield_paths.module b/modules/contrib/filefield_paths/filefield_paths.module
index 51e4d1f..a01c3c1 100644
--- a/modules/contrib/filefield_paths/filefield_paths.module
+++ b/modules/contrib/filefield_paths/filefield_paths.module
@@ -356,6 +356,12 @@ function filefield_paths_entity_update(EntityInterface $entity) {
         if ($definition instanceof ThirdPartySettingsInterface) {
           $settings = $definition->getThirdPartySettings('filefield_paths');
           if (isset($settings['enabled']) && $settings['enabled']) {
+            $destination = \Drupal::config('filefield_paths.settings')->get('temp_location');
+            foreach($field->referencedEntities() as $file) {
+              if(!file_exists($destination.'/'.$file->getFilename())) {
+                $new_file = file_move($file, $destination.'/'.$file->getFilename(), FILE_EXISTS_REPLACE);
+              }
+            }
             // Invoke hook_filefield_paths_process_file().
             foreach (\Drupal::moduleHandler()
                        ->getImplementations('filefield_paths_process_file') as $module) {