diff --git a/core/modules/views/src/Form/ViewsExposedForm.php b/core/modules/views/src/Form/ViewsExposedForm.php
index bbeb4db276..95ddaed5b1 100644
--- a/core/modules/views/src/Form/ViewsExposedForm.php
+++ b/core/modules/views/src/Form/ViewsExposedForm.php
@@ -81,6 +81,11 @@ public function buildForm(array $form, FormStateInterface $form_state) {
     /** @var \Drupal\views\ViewExecutable $view */
     $view = $form_state->get('view');
     $display = &$form_state->get('display');
+    // Existing arguments need to be passed as this exposed form might
+    // be used in a block. Without this contextual views arguments
+    // will be lost.
+    $args = $this->buildArgs();
+    $view->setArguments($args);
 
     $form_state->setUserInput($view->getExposedInput());
 
@@ -128,6 +133,17 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#id' => Html::getUniqueId('edit-submit-' . $view->storage->id()),
     ];
 
+    if ($view->hasUrl()) {
+      foreach($view->getUrl()->getRouteParameters() as $k => $parameter){
+        if($parameter != 'all'){
+          $newParameters[$k] = $parameter;
+        }
+      }
+      if(isset($newParameters)){
+        $view_url = $view->getUrl()->setRouteParameters($newParameters);
+      }
+    }
+
     if (!$view->hasUrl()) {
       // On any non views.ajax route, use the current route for the form action.
       if ($this->getRouteMatch()->getRouteName() !== 'views.ajax') {
@@ -138,6 +154,9 @@ public function buildForm(array $form, FormStateInterface $form_state) {
         $form_action = Url::fromUserInput($this->currentPathStack->getPath())->toString();
       }
     }
+    elseif(isset($view_url)){
+      $form_action = $view_url->toString();
+    }
     else {
       $form_action = $view->getUrl()->toString();
     }
@@ -156,6 +175,44 @@ public function buildForm(array $form, FormStateInterface $form_state) {
     return $form;
   }
 
+  /**
+   * @return array
+   *
+   * This code copied from \Drupal\views\Routing\ViewPageController::handle
+   *
+   * TODO: Don't repeat it.
+   *
+   * @see \Drupal\views\Routing\ViewPageController::handle
+   * @see https://www.drupal.org/project/drupal/issues/2821962
+   */
+  protected function buildArgs() {
+    /** @var \Drupal\Core\Routing\CurrentRouteMatch $route_match */
+    $route_match = \Drupal::service('current_route_match');
+    $args = [];
+    $route = $route_match->getRouteObject();
+    $map = $route->hasOption('_view_argument_map') ? $route->getOption('_view_argument_map') : [];
+
+    foreach ($map as $attribute => $parameter_name) {
+      // Allow parameters be pulled from the request.
+      // The map stores the actual name of the parameter in the request. Views
+      // which override existing controller, use for example 'node' instead of
+      // arg_nid as name.
+      if (isset($map[$attribute])) {
+        $attribute = $map[$attribute];
+      }
+      
+      if(!isset($arg) || $arg != $route_match->getRawParameter($attribute)) {
+        $arg = $route_match->getParameter($attribute);
+      }
+      
+      if (isset($arg) && $arg != 'all') {
+        $args[] = $arg;
+      }
+    }
+
+    return $args;
+  }
+
   /**
    * {@inheritdoc}
    */
